### PR TITLE
Refine package structure

### PR DESCRIPTION
**Pull request by luoxiu**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -4,32 +4,24 @@ branches:
...
       dist: xenial
       sudo: required
       env: 
         - SWIFT_VERSION="5.1"
-        - COMBINEX_EXPERIMENTAL_OBSERVABLE_OBJECT = YES
-    - os: osx
-      language: swift
-      osx_image: xcode11.1
-      env: SWIFT_PACKAGE_USE_COMBINEX="YES"
+        - CX_EXPERIMENTAL_OBSERVABLE_OBJECT="YES"
```
Comment (ddddxxx) [inline .travis.yml:19]: So we don't test with experimental features disabled? even if most users use CombineX this way.

We could accidentally use experimental components in normal code without breaking CI.

File: DevOpen [hunk 2]
```diff
@@ -3,7 +3,8 @@
 # Open the project in Xcode with env variables.
 # Enable all experimental features and test against CombineX.
 
-export SWIFT_PACKAGE_USE_COMBINEX="true"
-export COMBINEX_EXPERIMENTAL_OBSERVABLE_OBJECT="true"
+
+export CX_IMPLEMENTATION="CombineX"
```
Comment (ddddxxx) [inline DevOpen]: CX_IMPLEMENTATION seems too generic. probably CX_COMBINE_IMPLEMENTATION or CX_UNDERLAYING_COMBINE?

File: Package.swift [hunk 3]
```diff
@@ -3,183 +3,188 @@
 import Foundation
 import PackageDescription
 
-enum CombineImplementation: CaseIterable {
+// MARK: Helpers
+let env = ProcessInfo.processInfo.environment
+
+extension Optional where Wrapped: RangeReplaceableCollection {
```
Comment (ddddxxx) [inline Package.swift]: Put these helpers behind package declaration?

File: Package.swift [hunk 4]
```diff
@@ -3,183 +3,188 @@
...
-            swiftSettings: testSwiftSettings),
-    ],
-    swiftLanguageVersions: [
-        .v5
-    ]
-)
+let currentCombineImp = selectCombineImp()
+currentCombineImp.configure(package)
 
+// Pass the swift settings of current combine implementation and all experimental features to all test targets.
```
Comment (ddddxxx) [inline Package.swift]: I think experimental features should affect test target as well. These test should pass normally or fail as expected.

File: Package.swift [hunk 5]
```diff
@@ -3,183 +3,188 @@
...
     }
 }
 
-func configExperimentalFeatures() -> ExperimentalFeature {
-    let env = ProcessInfo.processInfo.environment
-    let featureSet = ExperimentalFeature.allCases.filter { env.keys.contains($0.environmentFlag!) }
-    return ExperimentalFeature(featureSet)
+func selectCombineImp() -> CombineImplementation {
+    let imp = env["CX_IMPLEMENTATION"]?.lowercased()
+    switch imp {
```
Comment (ddddxxx) [inline Package.swift:164]: It might be better to create `CombineImplementation` from string value.

File: Package.swift [hunk 6]
```diff
@@ -3,183 +3,188 @@
...
+        return .init(self.allCases)
+    }
     
-    var environmentFlag: String {
+    var isEnabled: Bool {
         switch self {
-        case .combine: return "SWIFT_PACKAGE_USE_COMBINE"
-        case .combineX: return "SWIFT_PACKAGE_USE_COMBINEX"
-        case .openCombine: return "SWIFT_PACKAGE_USE_OPEN_COMBINE"
+        case .observableObject: return env["CX_EXPERIMENTAL_OBSERVABLE_OBJECT"] != nil
```
Comment (ddddxxx) [inline Package.swift]: I think `CXExperimentalFeatures` should be irrelevant to environment variable.

File: Package.swift [hunk 7]
```diff
@@ -3,183 +3,188 @@
...
     
-    var packageDependencies: [Package.Dependency] {
-        switch self {
-        case .combine, .combineX:
-            return []
-        case .openCombine:
-            return [
-                .package(url: "https://github.com/broadwaylamb/OpenCombine", .branch("master")),
-            ]
+    static var enabled: CXExperimentalFeatures {
```
Comment (ddddxxx) [inline Package.swift]: It's more suitable for `ProcessInfo` than `CXExperimentalFeatures`.

**Review by ddddxxx — CHANGES_REQUESTED** [review]
Great! It's a huge readability improvement for Package.swift. and I find a few place could be improved more :)

File: Package.swift [hunk 8]
```diff
@@ -3,183 +3,188 @@
...
+        .target(name: "CXLibc"),
+        .target(name: "CXUtility"),
+        .target(name: "CXNamespace"),
+        .target(name: "CombineX", dependencies: ["CXLibc", "CXUtility", "CXNamespace"]),
+        .target(name: "CXFoundation", dependencies: ["CXUtility", "CXNamespace", "CombineX"]),
+        .target(name: "CXCompatible", dependencies: ["CXNamespace"]),
+        .target(name: "CXShim", dependencies: [/* depends on combine implementation */]),
+        .testTarget(name: "CombineXTests", dependencies: ["CXUtility", "CXShim", "Quick", "Nimble"]),
+        .testTarget(name: "CXFoundationTests", dependencies: ["CXShim", "Quick", "Nimble"]),
+        .testTarget(name: "CombineXFailingTests", dependencies: ["CXShim", "Quick", "Nimble"]),
```
Comment (ddddxxx) [inline Package.swift]: `CombineXFailingTests` or `CXFailingTests`?

This target should not be tested by CI.

File: .travis.yml (hunk 1 above)
Comment (luoxiu) [reply .travis.yml:19]: @ddddxxx Wow, I didn't think of this. I will bring it back.

**Review by ddddxxx — APPROVED** [review]

**Pull request merged**